plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/inspections/collections/SimplifiableCallChainInspection.kt

```diff
@@ -97,25 +101,48 @@ class SimplifiableCallChainInspection : AbstractCallChainChecker() {
         })
     }
 
-    private fun ResolvedCall<*>.lastFunctionalArgumentReturnType(context: BindingContext): KotlinType? {
+    private class FunctionalArgument(val isLambda: Boolean, val returnType: KotlinType?, val lastStatement: KtExpression?)
+
+    private fun ResolvedCall<*>.lastFunctionalArgument(context: BindingContext): FunctionalArgument? {
         val argument = valueArguments.entries.lastOrNull()?.value?.arguments?.firstOrNull()
         return when (val argumentExpression = argument?.getArgumentExpression()) {
             is KtLambdaExpression -> {
-                val functionLiteral = argumentExpression.functionLiteral
-                val body = argumentExpression.bodyExpression
-                val lastStatementType = body?.statements?.lastOrNull()?.getType(context)
-                val returnedTypes = body
-                    ?.collectDescendantsOfType<KtReturnExpression> { it.getTargetFunction(context) == functionLiteral }
-                    ?.mapNotNull { it.returnedExpression?.getType(context) }
-                    .orEmpty()
-                val types = listOfNotNull(lastStatementType) + returnedTypes
-                types.firstOrNull { it.isNullable() } ?: types.firstOrNull()
+                val lastStatement = argumentExpression.bodyExpression?.lastBlockStatementOrThis()
+                FunctionalArgument(
+                    isLambda = true,
+                    returnType = lastStatement?.getType(context),
+                    lastStatement = lastStatement
+                )
```

Comment:
ReturnExpression is checked by `AbstractCallChainChecker#findQualifiedConversion`, so no need to check it here.

https://github.com/JetBrains/intellij-community/blob/4430b3bfe94a48bd1982f74b331da39c36e6373d/plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/inspections/collections/AbstractCallChainChecker.kt#L47-L49